tests/test_known_flatland_bugs.py

```diff
@@ -0,0 +1,992 @@
+    """
+
+    env = init_test_rail_env(1)
+    env.step({0: RailEnvActions.DO_NOTHING, 1: RailEnvActions.DO_NOTHING})
+
+    agent = env.agents[0]
+    assert agent.state == TrainState.READY_TO_DEPART
+    min_distance_off_map = env.distance_map.get()[
+        agent.handle, agent.initial_position[0], agent.initial_position[1], agent.direction
+    ]
+    off_map_position = agent.initial_position
+
+    env.step({0: RailEnvActions.MOVE_FORWARD, 1: RailEnvActions.MOVE_FORWARD})
+    assert agent.state == TrainState.MOVING
+    min_distance_on_map = env.distance_map.get()[
+        agent.handle, agent.initial_position[0], agent.initial_position[1], agent.direction
+    ]
+    on_map_position = agent.position
+    assert np.all(on_map_position == off_map_position)
+    assert min_distance_off_map == min_distance_on_map
+
+
+def test_min_distance_for_off_map_trains_speed_of_half() -> None:
+    """The minimum distance for an off-map train is calculated from the initial position to the target. However, in
+    order for the agent to spawn or be placed on map on the initial position one action is needed.
+    As such the minimum distance can be viewed as being 1-off for off-map trains when it comes to the number of steps
+    needed to reach the target and especially when reasoning on whether the train can reach its target in time.
+    Although not strictly a bug, but something to be still aware of.
+    """
+    rail_env = init_test_rail_env(0.5)
+
+    rail_env.step({0: RailEnvActions.DO_NOTHING, 1: RailEnvActions.DO_NOTHING})
+    rail_env.step({0: RailEnvActions.DO_NOTHING, 1: RailEnvActions.DO_NOTHING})
+
+    agent = rail_env.agents[0]
+    assert agent.state == TrainState.READY_TO_DEPART
+    min_distance_off_map = rail_env.distance_map.get()[
+        agent.handle, agent.initial_position[0], agent.initial_position[1], agent.direction
+    ]
+    off_map_position = agent.initial_position
+
+    rail_env.step({0: RailEnvActions.MOVE_FORWARD, 1: RailEnvActions.MOVE_FORWARD})
+    assert agent.state == TrainState.MOVING
+    min_distance_on_map = rail_env.distance_map.get()[
+        agent.handle, agent.initial_position[0], agent.initial_position[1], agent.direction
+    ]
+    on_map_position = agent.position
+    assert np.all(on_map_position == off_map_position)
+    assert min_distance_off_map == min_distance_on_map
+
+
+# pylint: disable=protected-access
+def test_earliest_departure_zero_bug_BYDESIGN() -> None:
+    """
+    TODO revise design: by design of https://flatland-association.github.io/flatland-book/environment/environment/agent.html#state-machine,
+         an agent can go from WAITING to READY_TO_DEPART only after the first step transition. However, the design may be questioned.
+
+    Trains that have the earliest departure at ts 0 cannot be dispatched at ts 0 but only at ts 1. It seems like
+    every train starts with train state Waiting no matter the earliest departure.
+    """
```

Comment:
@castagna-a I think this is rather by design of [state machine](https://flatland-association.github.io/flatland-book/environment/environment/agent.html#state-machine) than a bug, an agent can go from `WAITING` to `READY_TO_DEPART` only after the first step transition. However, I agree the design can be questioned.